proposals/partial-type-inference.md

```diff
@@ -0,0 +1,738 @@
+
+## Summary
+[summary]: #summary
+
+Partial type inference introduces a syntax skipping obvious type arguments in the argument list of
+
+1. *invocation_expression*
+2. *object_creation_expresssion*
+
+and allowing to specify just ambiguous ones.
+
+> Example of skipping obvious type arguments
+>
+> ```csharp
+> M<_, object>(42, null); 
+> void M<T1, T2>(T1 t1, T2 t2) { ... }
+> ```
+
+It also improves the type inference in the case of *object_creation_expression* by leveraging type bounds obtained from the target, *object_or_collection_initializer*, and *type_parameter_constraints_clauses*. 
+
+> Example of inferring type arguments of type from the initializer list
+> 
+> ```csharp
+> using System.Collections.Generic;
+> var statistics = new Dictionary<,>(){["Joe"] = 20};
+> ```
+
+Besides the changes described above, the proposal mentions further interactions and possibilities to extend the partial type inference.
+
+## Motivation
+[motivation]: #motivation
+
+* The current method type inference has two possible areas for improvement.
+  The first improvement regards the strength of the method type inference, which uses only arguments' types to deduce the method's type arguments.
+  Concretely, we can see the weakness in cases where type arguments only depend on target type or type parameter restrictions.
+
+  > Example
+  >
+  > ```csharp
+  > object data = database.fetch();
+  > int count = data.Field("count"); // Error, method type inference can't infer the return type. We have to specify the type argument.
+  > 
+  > public static class Extensions
+  > {
+  >     public static TReturn Field<TReturn>(this object inst, string fieldName) { ... }
+  > }
+  > ```
+  >
+  > ```csharp
+  > test(new MyData()); // Error, method type inference can't infer T. We have to specify all type arguments.
+  >
+  > public void test<T, U>(U data) where T : TestCaseDefault<U> { ... }
+  > ```
+
+  The second improvement regards the "all or nothing" principle, where the method type inference infers either all of the type arguments or nothing.
+
+  > Example
+  > ```csharp
+  > log<, object>(new Message( ... ), null); // Error, impossible to specify just one type argument. We have to specify all of them. 
+  >
+  > public void log<T, U>(T message, U appendix) { ... }
+  > ```
+
+  The first improvement, which would improve the method type inference algorithm, has a significant disadvantage of the breaking change.
+  On the other hand, the second improvement, which would enable specifying some of the method's type arguments, does not influence old code, solves problems regarding the "all or nothing" principle, and reduces the first weakness.
```

Comment:
I'm unclear what the "all or nothing" principle means here.

Reply:
Line [58](https://github.com/dotnet/csharplang/pull/7582/files#diff-98ed61744ec680992e9779b8a70429db7e853f6942a0cb4e94b43b8e7ae96b91R58) explains "all or nothing" principle. The method type inference infers either all of the type arguments or nothing (e.g. we can't hint the compiler some of the type arguments).

Reply:
In line [68](https://github.com/dotnet/csharplang/pull/7582/files/7aa277e63f69c12f3bd27cae6c98a7a965fa6958#diff-98ed61744ec680992e9779b8a70429db7e853f6942a0cb4e94b43b8e7ae96b91R68), wouldn't it even be possible to infer twice: `test<TestCaseDefault<_>, _>(new MyData());` ?